language-examples/remedial-js/9.0-ES5-vs-ES6/README.md

```diff
@@ -0,0 +1,943 @@
+
+// Expression interpolation. One use is readable inline math.
+var a = 50;
+var b = 100;
+console.log(`The number of JS frameworks is ${a + b} and not ${2 * a + b}.`);
+
+// Multi-line strings without needing \n\
+console.log(`string text line 1
+string text line 2`);
+
+// Functions inside expressions
+function fn() { return 'result'; }
+console.log(`foo ${fn()} bar`);
+```
+
+ES5:
+
+```js
+'use strict';
+
+// Basic usage with an expression placeholder
+var person = 'Addy Osmani';
+console.log('Yo! My name is ' + person + '!');
+
+// Expressions work just as well with object literals
+var user = { name: 'Caitlin Potter' };
+console.log('Thanks for getting this into V8, ' + user.name + '.');
+
+// Expression interpolation. One use is readable inline math.
+var a = 50;
+var b = 100;
+console.log('The number of JS frameworks is ' + (a + b) + ' and not ' + (2 * a + b) + '.');
+
+// Multi-line strings:
+console.log('string text line 1\nstring text line 2');
+// Or, alternatively:
+console.log('string text line 1\n\
+string text line 2');
+
+// Functions inside expressions
+function fn() {
+  return 'result';
+}
+console.log('foo ' + fn() + ' bar');
+```
+
+
+## Computed Property Names
+
+Computed property names allow you to specify properties in object literals based on expressions:
+
+ES6:
+
+```js
+var prefix = 'foo';
+var myObject = {
+  [prefix + 'bar']: 'hello',
+  [prefix + 'baz']: 'world'
+};
+
+console.log(myObject['foobar']);
+// -> hello
+console.log(myObject['foobaz']);
+// -> world
+```
+
+ES5:
+
+```js
+'use strict';
+
+var prefix = 'foo';
+var myObject = {};
+
+myObject[prefix + 'bar'] = 'hello';
+myObject[prefix + 'baz'] = 'world';
+
+console.log(myObject['foobar']);
+// -> hello
+console.log(myObject['foobaz']);
+// -> world
+```
+
+## Destructuring Assignment
+
+The destructuring assignment syntax is a JavaScript expression that makes it possible to extract data from arrays or objects using a syntax that mirrors the construction of array and object literals.
+
+
+ES6:
+
+```js
+var {foo, bar} = {foo: 'lorem', bar: 'ipsum'};
+// foo => lorem and bar => ipsum
+```
+
+ES5:
+
+```js
+'use strict';
+
+var _ref = { foo: 'lorem', bar: 'ipsum' };
+
+// foo => lorem and bar => ipsum
+var foo = _ref.foo;
+var bar = _ref.bar;
+```
+
+ES3:
+
+```js
+with({foo: 'lorem', bar: 'ipsum'}) {
+  // foo => lorem and bar => ipsum
+}
+```
+
+ES6:
+
+```js
+var [a, , b] = [1,2,3];
+```
+
+ES6 (shimming using `Symbol.iterator`):
+
+```js
+'use strict';
+
+var _slicedToArray = function (arr, i) {
+  if (Array.isArray(arr)) {
+    return arr;
+  } else {
+    var _arr = [];
+
+    for (var _iterator = arr[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
+      _arr.push(_step.value);
+
+      if (i && _arr.length === i) {
+        break;
+      }
+    }
+
+    return _arr;
+  }
+};
+
+var _ref = [1, 2, 3];
+
+var _ref2 = _slicedToArray(_ref, 3);
+
+var a = _ref2[0];
+var b = _ref2[2];
```

Comment:
Might take this section out, as its more about Symbol.iterator as a shim, which I don't know of a good use case for.